src/core/manager.c

```diff
@@ -2285,16 +2285,35 @@ static int manager_dispatch_cgroups_agent_fd(sd_event_source *source, int fd, ui
 }
 
 static bool manager_process_barrier_fd(const char *buf, FDSet *fds) {
+        const char *p;
+
         assert(buf);
 
         /* nothing else must be sent when using BARRIER=1 */
-        if (STR_IN_SET(buf, "BARRIER=1", "BARRIER=1\n")) {
-                if (fdset_size(fds) != 1)
-                        log_warning("Got incorrect number of fds with BARRIER=1, closing them.");
-                return true;
-        } else if (startswith(buf, "BARRIER=1\n") || strstr(buf, "\nBARRIER=1\n") || endswith(buf, "\nBARRIER=1"))
-                log_warning("Extra notification messages sent with BARRIER=1, ignoring everything.");
+        p = strstr(buf, "BARRIER=1");
+        if (p != NULL) {
+                int barrier_found = -1;
+
+                if (p == buf) {
+                        /* Check there is nothing after BARRIER=1 */
+                        p += STRLEN("BARRIER=1");
+                        if ((p[0] == '\0') || (p[0] == '\n' && p[1] == '\0'))
+                                barrier_found = 1;
+                } else {
+                        /* Check BARRIER=1 is not at the beginning of a new line */
+                        if (p[-1] != '\n')
+                                barrier_found = 0;
+                }
+
+                if (barrier_found > 0) {
+                        if (fdset_size(fds) != 1)
+                                log_warning("Got incorrect number of fds with BARRIER=1, closing them.");
+                } else if (barrier_found < 0)
+                        log_warning("Extra notification messages sent with BARRIER=1, ignoring everything.");
 
+                /* Drop the message if BARRIER=1 was found */
+                return (barrier_found != 0);
+        }
```

Comment:
this isn't right. if I send FOOBARRIER=1000, then this will generate an invalid warning. And if i send FOOBARRIER=100\nBARRIER=1\n it will also generate an invalid warning, though a different, also wrong one.

Also, I doubt it's worth it. the old code in the worst case did four str[n]cmp(), and two str[str|chr](). Of these one could argue that one could optimize maybe one str[n]cmp() away, plus maybe one str[str|chr](). The former shouldn't matter at all, and the letter likely not either, because if today's CPU are quick at something it is comparing and copying memory. It's very very hard to beat that, and additional if checks are usually worse than comparing/copying a few bytes too many...

Can you show that this matters? i.e. in a profiling run?

Reply:
No it does not matter much. And you are right there are some flow in this patch, but 
- `FOOBARRIER=1000` should not generate a warning (I missed something ?)
- `BARRIER=1000` is going to be handled like `BARRIER=1` but I think it was fine
- `FOOBARRIER=100\nBARRIER=1\n` is indeed not properly handled, is it not going to generate a warning (we do not see that there is `BARRIER=1` in it), and it must generate one

And yes this does not really matters, This is negligible (I have done it since I think it could improve it), but it is not improving anything... So I think we can drop this patch except that ...

This function should should return true if `BARRIER=1` is found in one "line" in the message, right ? Currently this is not the case.